atomicfu-transformer/src/main/kotlin/kotlinx/atomicfu/transformer/AtomicFUTransformer.kt

```diff
@@ -806,46 +800,126 @@ class AtomicFUTransformer(
                 is MethodInsnNode -> {
                     // invoked virtual method on atomic var -- fixup & done with it
-                    debug("invoke $f.${j.name}", sourceInfo.copy(i = j))
-                    return fixupInvokeVirtual(ld, false, j, f)
+                    debug("invoke $f.${operation.name}", sourceInfo.copy(i = operation))
+                    return fixupInvokeVirtual(ld, arrayElementInfo != null, operation, f)
                 }
                 is VarInsnNode -> {
+                    val onArrayElement = arrayElementInfo != null
+                    check((f.getPrimitiveType(vh).sort == ARRAY) == (arrayElementInfo != null))
                     // was stored to local -- needs more processing:
                     // store owner ref into the variable instead
-                    val v = j.`var`
-                    val next = j.next
-                    instructions.remove(ld)
-                    val lv = localVar(v, j)
+                    val v = operation.`var`
+                    val next = operation.next
+                    if (onArrayElement) {
+                        // leave just owner class load insn on stack
+                        arrayElementInfo!!.forEach { instructions.remove(it) }
+                    } else {
+                        instructions.remove(ld)
+                    }
+                    val lv = localVar(v, operation)
                     if (lv != null) {
                         // Stored to a local variable with an entry in LVT (typically because of inline function)
-                        if (lv.desc != f.fieldType.descriptor)
+                        if (lv.desc != f.fieldType.descriptor && !onArrayElement)
                             abort("field $f was stored to a local variable #$v \"${lv.name}\" with unexpected type: ${lv.desc}")
                         // correct local variable descriptor
                         lv.desc = f.ownerType.descriptor
                         lv.signature = null
                         // process all loads of this variable in the corresponding local variable range
                         forVarLoads(v, lv.start, lv.end) { otherLd ->
-                            fixupVarLoad(f, ld, otherLd)
+                            fixupLoad(f, ld, otherLd, arrayElementInfo)
                         }
                     } else {
                         // Spilled temporarily to a local variable w/o an entry in LVT -> fixup only one load
-                        fixupVarLoad(f, ld, nextVarLoad(v, next))
+                        fixupLoad(f, ld, nextVarLoad(v, next), arrayElementInfo)
                     }
                     return next
                 }
                 else -> abort("cannot happen")
             }
         }
 
+        private fun fixupLoad(f: FieldInfo, ld: FieldInsnNode, otherLd: VarInsnNode, arrayElementInfo: List<AbstractInsnNode>?): AbstractInsnNode? {
+            return if (arrayElementInfo != null) {
```

Comment:
Style nit: This better looks as expression function (can be fixed with Alt+ENTER -> Convert to expression body:
```
private fun fixupLoad(...) = 
    if (....) ....
```